picocli-shell-jline3/src/main/java/picocli/tool/jline3/ExampleTool.java

```diff
@@ -0,0 +1,146 @@
+package picocli.tool.jline3;
+
+import org.jline.reader.EndOfFileException;
+import org.jline.reader.LineReader;
+import org.jline.reader.LineReaderBuilder;
+import org.jline.reader.UserInterruptException;
+import org.jline.terminal.Terminal;
+import org.jline.terminal.TerminalBuilder;
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Parameters;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+import static java.lang.System.exit;
+import static java.util.Collections.emptyList;
+import static picocli.tool.jline3.Repl.runRepl;
+import static picocli.tool.jline3.YourProcessing.process;
+
+/**
+ * "ExampleTool" demonstrates a command-line tool using Picocli and JLine3
+ * together that follow standard UNIX/Linux practices: <ul>
+ * <li>Provide colorful help or version based on command-line flags (no
+ * further processing)</li>
+ * <li>Process command-line as program input</li>
+ * <li>Process <code>STDIN</code> (a pipe, shell redirection, etc) as
+ * program input</li>
+ * <li>Provide a REPL (shell) to process input interactively including
+ * expected editing features at a prompt</li>
+ * </ul>
+ */
+public class ExampleTool {
+    public static void main(String[] args) {
+        exit(new CommandLine(new Options()).execute(args));
+    }
+}
+
+@Command(name = "example tool", mixinStandardHelpOptions = true, version = {"example tool 0-SNAPSHOT"})
+final class Options implements Callable<Integer> {
+    @Option(names = {"--prompt"}, description = {"Change the interactive prompt from '> '."})
+    private String prompt = "> ";
+
+    @Parameters(description = {"Command line arguments", "Edit this help as suitable."})
+    private List<String> arguments = emptyList();
+
+    @Override
+    public Integer call() {
+        // Return failure to demonstrate that `-h` does not fail: standard
+        // options are handled by Picocli: this `call()` method is not
+        // invoked by standard options
+
+        // Use case #1 -- process command line arguments
+        if (!arguments.isEmpty()) {
+            for (final String argument : arguments) {
+                final Integer result = process(argument);
+                if (0 != result) return result;
+            }
+            return 0;
+        }
+        // Use case #2 -- pipe or shell redirection
+        else if (null == System.console()) return runFromInput();
+        // Use case #3 -- interactive REPL
+        else return runRepl(prompt);
+    }
+
+    /** @todo Nicer way using more modern Java. */
+    private static Integer runFromInput() {
+        try (final BufferedReader lineReader = lineReader()) {
```

Comment:
This method seems to be a rudimentary re-implementation of JLine functionality.
I don't think we should suggest to users that they do this.
My recommendation to users would be to familiarize themselves with what JLine offers and use that instead.